src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.EasyRequest.cs

```diff
@@ -384,7 +399,53 @@ private void SetVerb()
+
+                    // For most content types and most HTTP methods, we use a callback that lets libcurl
+                    // get data from us if/when it wants it.  However, as an optimization, for POSTs that
+                    // use content already known to be entirely in memory, we hand that data off to libcurl
+                    // ahead of time.  This not only saves on costs associated with all of the async transfer
+                    // between the content and libcurl, it also lets libcurl do larger writes that can, for
+                    // example, enable fewer packets to be sent on the wire.
+                    var inMemContent = _requestMessage.Content as ByteArrayContent;
+                    ArraySegment<byte> contentSegment;
+                    if (inMemContent != null &&
+                        inMemContent.TryGetBuffer(out contentSegment) &&
+                        contentSegment.Count <= InMemoryPostContentLimit) // skip if we'd be forcing libcurl to allocate/copy a large buffer
+                    {
+                        // Only pre-provide the content if the content still has its ContentLength
+                        // and if that length matches the segment.  If it doesn't, something has been overridden,
+                        // and we should rely on reading from the content stream to get the data.
+                        long? contentLength = inMemContent.Headers.ContentLength;
+                        if (contentLength.HasValue && contentLength.GetValueOrDefault() == contentSegment.Count)
+                        {
+                            _inMemoryPostContent = true;
+
+                            // Debug double-check array segment; this should all have been validated by the ByteArrayContent
+                            Debug.Assert(contentSegment.Array != null, "Expected non-null byte content array");
+                            Debug.Assert(contentSegment.Count >= 0, $"Expected non-negative byte content count {contentSegment.Count}");
+                            Debug.Assert(contentSegment.Offset >= 0, $"Expected non-negative byte content offset {contentSegment.Offset}");
+                            Debug.Assert(contentSegment.Array.Length - contentSegment.Offset >= contentSegment.Count,
+                                $"Expected offset {contentSegment.Offset} + count {contentSegment.Count} to be within array length {contentSegment.Array.Length}");
+
+                            // Hand the data off to libcurl with COPYPOSTFIELDS for it to copy out the data. (The alternative
+                            // is to use POSTFIELDS, which would mean we'd need to pin the array in the ByteArrayContent for the
+                            // duration of the request.  Often with a ByteArrayContent, the data will be small and the copy cheap.)
+                            unsafe
+                            {
+                                fixed (byte* inMemContentPtr = contentSegment.Array)
```

Comment:
Took a look at the IL.  For:
```C#
fixed (byte* ptr = arr)
```
the compiler needs to account for the case that arr may be null or that arr may be empty, in which case it needs to assign ptr to null, so it essentially needs to generate code equivalent to:
```C#
ptr = arr != null && arr.Length > 0 ? &arr[0] : null;
```
In contrast by writing:
```
fixed (byte* ptr = &arr[0])
```
it no longer needs those branches, since you'll get a null ref by attempting to dereference the array and an out of range error if the array is empty.

So there is actually a functional difference between the cases and C# can't just emit the same code for both (unless it could prove statically that the array wasn't null or empty).